File: metodi-di-pagamento-alternativi/googlepay/java/googlepay.java
                                            
// Google Pay - Integrazione tramite API

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.util.EntityUtils;
import org.json.JSONObject;

import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.Date;

public class googlepay {

    public static void main(String[] a) throws Exception {

        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMddHHmmss");
        Date date = new Date();

        // Parametri per calcolo MAC
        String apiKey = "<ALIAS>"; // Alias fornito da Nexi
        String chiaveSegreta = "<CHIAVE SEGRETA PER CALCOLO MAC>"; // Chiave segreta fornita da Nexi

        String requestUrl = "https://int-ecommerce.nexi.it/" + "ecomm/api/paga/googlePay";

        String codTrans = "TESTAP_" + dateFormat; // Codice della transazione
        String importo = "5000"; // 5000 = 50,00 EURO (indicare la cifra in centesimi)
        String divisa = "978"; // divisa 978 indica EUR
        String timeStamp = "" + System.currentTimeMillis();
        String jsonGoogle = "{}";

        String stringaMac = "apiKey=" + apiKey
          + "codiceTransazione=" + codTrans
          + "importo=" + importo
          + "divisa=" + divisa
          + "timeStamp=" + timeStamp
          + chiaveSegreta;

        // Calcolo MAC
        String macCalculated = hashMac(stringaMac);

        // Parametri di invio
        JSONObject json = new JSONObject();
        json.put("apiKey", apiKey);
        json.put("codiceTransazione", codTrans);
        json.put("importo", importo);
        json.put("divisa", divisa);
        json.put("googlePay", jsonGoogle);
        json.put("timeStamp", timeStamp);
        json.put("mac", macCalculated);
        

        // Chiamata API
        CloseableHttpClient httpClient = HttpClientBuilder.create().build();
        HttpPost request = new HttpPost(requestUrl);
        StringEntity params = new StringEntity(json.toString());
        request.addHeader("content-type", "application/json");
        request.setEntity(params);
        HttpResponse response = httpClient.execute(request);
        HttpEntity entity = response.getEntity();
        String responseString = EntityUtils.toString(entity, "UTF-8");

        // Parametri di ritorno
        JSONObject responseObj = new JSONObject(responseString);

        System.out.print(responseObj.toString());
    }

    public static String hashMac(String stringaMac) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-1");
        byte[] in = digest.digest(stringaMac.getBytes("UTF-8"));

        final StringBuilder builder = new StringBuilder();
        for (byte b : in) {
            builder.append(String.format("%02x", b));
        }
        return builder.toString();
    }
}
                    
                

File: metodi-di-pagamento-alternativi/googlepay/php/googlepay.php
                                            
<?php

// Google Pay - Integrazione tramite API

$connection = curl_init();

if ($connection) {
    $apiKey = "<ALIAS>"; // Sostituire con il valore fornito da Nexi
    $chiaveSegreta = "<CHIAVESEGRETA>"; // Sostituire con il valore fornito da Nexi

    $requestUrl = "https://int-ecommerce.nexi.it/";
    $requestUri = "ecomm/api/paga/googlePay";

    $codTrans = "TESTAP_" . date('YmdHis'); // Codice della transazione
    $importo = "5000"; // 5000 = 50,00 EURO (indicare la cifra in centesimi)
    $divisa = "978"; // divisa 978 indica EUR
    $timeStamp = (time()) * 1000;

    // Calcolo MAC
    $mac = sha1('apiKey=' . $apiKey . 'codiceTransazione=' . $codTrans . 'importo=' . $importo . "divisa=" . $divisa . "timeStamp=" . $timeStamp . $chiaveSegreta);

    $parametri = array(
        'apiKey' => $apiKey,
        'codiceTransazione' => $codTrans,
        'importo' => $importo,
        'divisa' => $divisa,
        'googlePay' => $jsonGoogle,
        'timeStamp' => (string) $timeStamp,
        'mac' => $mac
    );

    curl_setopt($connection, CURLOPT_HTTPHEADER, array('Content-Type: application/json'));
    curl_setopt($connection, CURLOPT_URL, $requestUrl . $requestUri);
    curl_setopt($connection, CURLOPT_POST, 1);
    curl_setopt($connection, CURLOPT_POSTFIELDS, json_encode($parametri));
    curl_setopt($connection, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($connection, CURLINFO_HEADER_OUT, true);

    $json = curl_exec($connection);

    curl_close($connection);

    // Decodifico risposta
    $risposta = json_decode($json, true);

    // Controllo JSON di risposta
    if (json_last_error() === JSON_ERROR_NONE) {

        $MACrisposta = sha1('esito=' . $risposta['esito'] . 'idOperazione=' . $risposta['idOperazione'] . 'timeStamp=' . $risposta['timeStamp'] . $chiaveSegreta);

        // Controllo MAC di risposta
        if ($risposta['mac'] == $MACrisposta) {

            // Controllo esito
            if ($risposta['esito'] == 'OK') {
                echo 'La transazione ' . $codTrans . " è avvenuta con successo; codice autorizzazione: " . $risposta['codiceAutorizzazione'];
            } else {
                echo 'La transazione ' . $codTrans . " è stata rifiutata; descrizione errore: " . $risposta['errore']['messaggio'];
            }
        } else {
            echo 'Errore nel calcolo del MAC di risposta';
        }
    } else {
        echo 'Errore nella lettura del JSON di risposta';
    }
} else {
    echo "Errore curl";
}
                    
                